### PR TITLE
Added Kanban Support for Feature Widget

### PR DESCRIPTION
**Pull request by amp5208**
Currently, the Feature Widget does not support Kanban teams, nor does it support teams which have both Kanban and Scrum teams simultaneously.  With this change, Hygieia now includes both of these use cases, for both Jira and VersionOne.  Related issue can be found here:  https://github.com/capitalone/Hygieia/issues/505.

Hygieia will now flip through both Kanban and Scrum boards if both exist for a user, also allowing for a user to pause the view switching behavior.  Also, the Kanban view will not include a Done display.  Below I have included reference screen shots:

_Scrum iterations for a single project in Jira, where there are multiple iterations in the project:_
image

_Kanban view for the same project in Jira, which also has Scrum iterations within the project:_
image

_Note:_ Enabling the new Kanban feature for existing Hygieia implementations may require dumping existing feature collections in MongoDB, to be re-populated with the updated Jira/VersionOne Feature Collectors.

@rashisaxena :  Please validate my fork and ensure that these changes also work for your use case.  Please check this box when completed validation:

 - [ ] Customer validation complete

@tabladrum / @jdamick :  Please provide a code review and add inline comments.  I have already fetched from upstream master, but not before squashing my commits; all commits can be reviewed together using the following link:  https://github.com/capitalone/Hygieia/pull/606/files.  Please only fully accept the pull request once all check boxes on this issue are complete:

 - [ ] Internal QA sign-off

### CODE REVIEW

File: UI/src/app/dashboard/core/data-factories/feature-data.js [hunk 1]
```diff
@@ -8,99 +8,162 @@
...
-		
+		var agileType = {
+			kanban : "&agileType=kanban",
+			scrum : "&agileType=scrum",
+		};
+
 		var testTotal = 'http://localhost:8080/api/feature/estimates/total/Team:9933916?component=552e7b21c44a49be6261d54a';
 		var buildTotal = '/api/feature/estimates/total/';
-		
+
```
Comment (tabladrum) [inline UI/src/app/dashboard/core/data-factories/feature-data.js]: Don't rely on local api. See this from /app/dashboard/core/data-factories/build-data.js:
        var testDetailRoute = 'test-data/build_detail.json';
        var buildDetailRoute = '/api/build/';

Put test JSON in /test-data folder.

Comment (amp5208) [inline UI/src/app/dashboard/core/data-factories/feature-data.js]: I can add that.  Currently, I am not testing in this way, however; we are deploying and then running Cucumber functional tests; API is tested explicitly using JUnit integration tests

File: versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java [hunk 2]
```diff
@@ -8,53 +9,105 @@
...
+					Integer.valueOf(startDate.substring(5, 7)),
+					Integer.valueOf(startDate.substring(8, 10)));
+			endCalendar.set(Integer.valueOf(endDate.substring(0, 4)),
+					Integer.valueOf(endDate.substring(5, 7)),
+					Integer.valueOf(endDate.substring(8, 10)));
+			long diffMill = endCalendar.getTimeInMillis() - startCalendar.getTimeInMillis();
+			long diffDays = TimeUnit.DAYS.convert(diffMill, TimeUnit.MILLISECONDS);
+
+			if (diffDays <= 28) {
+				// Scrum-enough
```
Comment (tabladrum) [inline versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java]: Should '28' be in settings file?

File: versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java [hunk 3]
```diff
@@ -8,53 +9,105 @@
...
+	 * Evaluates whether a sprint length appears to be kanban or scrum
+	 * 
+	 * @param startDate
+	 *            The start date of a sprint in ISO format
+	 * @param endDate
+	 *            The end date of a sprint in ISO format
+	 * @return True indicates a scrum sprint; False indicates a Kanban sprint
+	 */
+	public boolean evaluateSprintLength(String startDate, String endDate) {
+		boolean sprintIndicator = false;
```
Comment (tabladrum) [inline versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java]: This looks very complicated for date manipulation - specially the startDate.substring operations and subsequent check isInteger. Should be refactored.

File: versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java (hunk 2 above)
Comment (amp5208) [inline versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java]: I actually forgot to un-magic this number 😄 .  I will make the modification.

File: versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java (hunk 3 above)
Comment (amp5208) [inline versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java]: Agreed - however, this is only showing as "modified" due to a style re-format.  The only new method here is evaluateSprintLength, which I created test cases for, proving its use:  https://github.com/amp5208/Hygieia/blob/master/versionone-feature-collector/src/test/java/com/capitalone/dashboard/util/DateUtilTest.java.

Please accept this change, with the knowledge that the rest of the VersionOne Feature Collector will be refactored (referenced here: https://github.com/capitalone/Hygieia/issues/352)

Comment (tabladrum) [inline versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java]: Actually I was referring to this code snippet... Sorry my comment was in the wrong place:
f ((!StringUtils.isAnyEmpty(startDate) && !StringUtils.isAnyEmpty(endDate))
 +				&& (this.isInteger(startDate.substring(0, 4), 10)
 +						&& this.isInteger(endDate.substring(0, 4), 10)
 +						&& this.isInteger(startDate.substring(8, 10), 10)
 +						&& this.isInteger(endDate.substring(8, 10), 10)
 +						&& this.isInteger(startDate.substring(8, 10), 10)
 +						&& this.isInteger(endDate.substring(8, 10), 10))) {
 +			startCalendar.set(Integer.valueOf(startDate.substring(0, 4)),
 +					Integer.valueOf(startDate.substring(5, 7)),
 +					Integer.valueOf(startDate.substring(8, 10)));
 +			endCalendar.set(Integer.valueOf(endDate.substring(0, 4)),
 +					Integer.valueOf(endDate.substring(5, 7)),
 +					Integer.valueOf(endDate.substring(8, 10)));
 +			long diffMill = endCalendar.getTimeInMillis() - startCalendar.getTimeInMillis();
 +			long diffDays = TimeUnit.DAYS.convert(diffMill, TimeUnit.MILLISECONDS);

Comment (amp5208) [inline versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java]: Ah yes, that section... I actually had a bit of trouble working with the stringified dates-to-date object model format here.  The Java built-in API for date manipulation is notably flawed _(source, one of many: http://stackoverflow.com/questions/1969442/whats-wrong-with-java-date-time-api)_.  To make this better, I need to pull in a different date library for use.

Which date library would you suggest I use for converting ISO date strings to date objects, and then finding the difference between the two dates as an integer value?

Comment (tabladrum) [inline versionone-feature-collector/src/main/java/com/capitalone/dashboard/util/DateUtil.java]: Yeah. I hate date time manipulations and formatting. I can help you rewriting this.

**Pull request merged**